reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscriber.scala

```diff
@@ -74,84 +83,135 @@ object StreamSubscriber {
+            (UpstreamCompletion: State).pure[F]
         }
         case OnError(e) => {
-          case WaitingOnUpstream(_, r) => UpstreamError(e) -> r.complete(e.asLeft)
-          case _                       => UpstreamError(e) -> F.unit
+          case WaitingOnUpstream(_) =>
+            (q.enqueue1(e.asLeft) >> q.enqueue1(None.asRight)).as(UpstreamError(e))
+          case Receiving(_) =>
+            q.enqueue1(None.asRight).as(UpstreamError(e))
+          case _ =>
+            (UpstreamError(e): State).pure[F]
         }
         case OnFinalize => {
-          case WaitingOnUpstream(sub, r) =>
-            DownstreamCancellation -> (F.delay(sub.cancel) >> r.complete(None.asRight))
-          case Idle(sub) => DownstreamCancellation -> F.delay(sub.cancel)
-          case o         => o -> F.unit
+          case WaitingOnUpstream(sub) =>
+            (F.delay(sub.cancel) >> q.enqueue1(None.asRight)).as(DownstreamCancellation)
+          case Idle(sub) =>
+            F.delay(sub.cancel).as(DownstreamCancellation)
+          case Receiving(sub) =>
+            F.delay(sub.cancel).as(DownstreamCancellation)
+          case o =>
+            o.pure[F]
         }
-        case OnDequeue(r) => {
-          case Uninitialized          => RequestBeforeSubscription(r) -> F.unit
-          case Idle(sub)              => WaitingOnUpstream(sub, r) -> F.delay(sub.request(1))
-          case err @ UpstreamError(e) => err -> r.complete(e.asLeft)
-          case UpstreamCompletion     => UpstreamCompletion -> r.complete(None.asRight)
-          case o                      => o -> r.complete((new Error(s"received request in invalid state [$o]")).asLeft)
+        case OnDequeue => {
+          case Uninitialized =>
+            (RequestBeforeSubscription: State).pure[F]
+          case Receiving(sub) =>
+            (Receiving(sub): State).pure[F]
+          case err @ UpstreamError(e) =>
+            (q.enqueue1(e.asLeft) >> q.enqueue1(None.asRight)).as(err)
+          case Idle(sub) =>
+            F.delay(sub.request(maxDemand)).as(WaitingOnUpstream(sub)) // request on first dequeue
+          case UpstreamCompletion =>
+            q.enqueue1(None.asRight).as(UpstreamCompletion)
+          case st => st.pure[F]
         }
       }
 
-    Ref.of[F, State](Uninitialized).map { ref =>
+    Ref.of[F, State](Uninitialized) map { ref =>
       new FSM[F, A] {
-        def nextState(in: Input): F[Unit] = ref.modify(step(in)).flatten
-        def onSubscribe(s: Subscription): F[Unit] = nextState(OnSubscribe(s))
-        def onNext(a: A): F[Unit] = nextState(OnNext(a))
-        def onError(t: Throwable): F[Unit] = nextState(OnError(t))
-        def onComplete: F[Unit] = nextState(OnComplete)
-        def onFinalize: F[Unit] = nextState(OnFinalize)
-        def dequeue1: F[Either[Throwable, Option[A]]] =
-          Deferred[F, Out].flatMap(p => ref.modify(step(OnDequeue(p))).flatten >> p.get)
+        def nextState(in: Input): F[Unit] =
+          ref.get
```

Comment:
This has different semantics to `modify` in the presence of concurrent access since two operations may see the same initial state and whichever finishes last will win rather than be retried.